[Model] support input embeddings for qwen2vl

whyiug opened the pull request:
FILL IN THE PR DESCRIPTION HERE

FIX #8857

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

whyiug committed:
support input embeddings for qwen2vl

DarkLight1337 commented:
@fyabc can you assist OP to address https://github.com/vllm-project/vllm/pull/6613#issuecomment-2377237573?

whyiug commented:
Adding to our usage scenarios.
We need to perform multiple VQA tasks (same image, different questions) using several Qwen2VL models(finetuned). In this case, I can use this feature to deploy image encoding and language model inference in different pods. This will save much time by avoiding redundant image downloading and image encoding for each language model inference.
@fyabc

whyiug committed:
support input embeddings for qwen2vl

whyiug committed:
fix tiny lint

whyiug committed:
fix tiny lint

whyiug commented:
Can vllm support this multmodal input format?
To support multiple images, videos, dynamic resolution, and M-RoPE in the future.
@ywang96   please have a look and consider it.

```python
    image_embeds = ……
    # mm_data['image'] = image_embeds
    mm_data['image'] = {
        "image_embeds": image_embeds,
        "image_grid_thw":  inputs["image_grid_thw"],
    }

    llm_inputs = {
        "prompt": prompt,
        "multi_modal_data": mm_data,
    }

    outputs = llm.generate([llm_inputs], sampling_params=sampling_params)
    generated_text = outputs[0].outputs[0].text
````

whyiug committed:
refactor _expand_pad_tokens function

whyiug committed:
refactor _expand_pad_tokens function

whyiug committed:
fix lint

whyiug committed:
fix lint

whyiug committed:
fix lint

DarkLight1337 commented on vllm/model_executor/models/qwen2_vl.py:
```diff
@@ -76,19 +76,31 @@
     image_grid_thw: torch.Tensor
-    """Shape: `(num_images, 3)`
-    
+    """Shape: `(num_images, 3)
```
```suggestion
    """Shape: `(num_images, 3)`
```

DarkLight1337 commented on vllm/model_executor/models/qwen2_vl.py:
```diff
@@ -775,62 +792,60 @@ def input_processor_for_qwen2_vl(ctx: InputContext,
     # Expand image pad tokens.
-    if image_inputs is not None:
-        image_indices = [
+    def expand_pad_tokens(inputs, token_id, make_batched_fn, data_type_key):
```
Can you extract this function to the global scope to avoid conflicting variable names with the outer function scope?

Also, please add type annotations for each parameter.

whyiug replied:
In fact, that's how I first implemented it, but image_processor, prompt_token_ids should be the same, and it would introduce too many parameters and impact the global scope.

DarkLight1337 reviewed (commented):
Thanks for implementing this! Some comments.

> Can vllm support this multmodal input format?
To support multiple images, videos, dynamic resolution, and M-RoPE in the future.

For now, I suggest you override the default input mapper (in `register_image_input_mapper`) to support it for this model specifically. To maintain a consistent API, the regular case of just inputting a tensor should still be supported.

DarkLight1337 commented:
You should also update the Supported Models page in the docs to indicate that the model supports image embeddings.

whyiug commented:
> Thanks for implementing this! Some comments.感谢您实施此操作！一些评论。
> 
> > Can vllm support this multmodal input format?vllm可以支持这种多模态输入格式吗？
> > To support multiple images, videos, dynamic resolution, and M-RoPE in the future.未来支持多图像、视频、动态分辨率和M-RoPE。
> 
> For now, I suggest you override the default input mapper (in `register_image_input_mapper`) to support it for this model specifically. To maintain a consistent API, the regular case of just inputting a tensor should still be supported.目前，我建议您覆盖默认输入映射器（在`register_image_input_mapper`中）以专门支持该模型。为了保持 API 的一致性，仍应支持仅输入张量的常规情况。

We have already modified the default input mapper for Qwen2VL (`image_input_mapper_for_qwen2_vl`). Are you suggesting that we create an additional mapper? I can incorporate logic such as `elif isinstance(data, torch.Tensor)` to support input tensors in `image_input_mapper_for_qwen2_vl`; however, please note that the results may be somewhat inconsistent with the previous implementation. This is due to the fact that Qwen2VL requires `image_grid_thw` for calculating M-RoPE.

Right now it only works on Qwen2VL and does not affect the inputs of other models.
What I'm thinking about is how vllm will come to expand the image embedding inputs in the future, as well as embedding inputs support for video.
By the way, I have implemented Qwen2VL for input video embedding, haven't submitted it.

whyiug committed:
Update the supported_models.rst file  for Qwen2-VL model.

DarkLight1337 commented:
> > For now, I suggest you override the default input mapper (in `register_image_input_mapper`) to support it for this model specifically. To maintain a consistent API, the regular case of just inputting a tensor should still be supported.目前，我建议您覆盖默认输入映射器（在`register_image_input_mapper`中）以专门支持该模型。为了保持 API 的一致性，仍应支持仅输入张量的常规情况。
> 
> We have already modified the default input mapper for Qwen2VL (`image_input_mapper_for_qwen2_vl`). Are you suggesting that we create an additional mapper? I can incorporate logic such as `elif isinstance(data, torch.Tensor)` to support input tensors in `image_input_mapper_for_qwen2_vl`; however, please note that the results may be somewhat inconsistent with the previous implementation. This is due to the fact that Qwen2VL requires `image_grid_thw` for calculating M-RoPE.

Sorry, I forgot there is already an existing input mapper. In that case, you can update the existing one.

whyiug committed:
Refactor _expand_pad_tokens function for Qwen2-VL model

whyiug committed:
Refactor _expand_pad_tokens function for Qwen2-VL model

whyiug committed:
Refactor typing imports in Qwen2-VL model

whyiug commented:
@DarkLight1337 This ci's failure seems to have nothing to do with my changes. please have a look.

DarkLight1337 commented:
You should merge in the changes from main branch to resolve the CI failures.

DarkLight1337 commented:
> You should also update the Supported Models page in the docs to indicate that the model supports image embeddings.

Since the input format is different for this model, I suggest also adding a note to the documentation to explain how to input embeddings for Qwen2-VL.

whyiug committed:
Merge remote-tracking branch 'origin' into feature_sup_qwen2vl_input_embedding

whyiug committed:
Refactor Qwen2-VL model to support inference with image embeddings and additional parameters

whyiug commented:
> > You should also update the Supported Models page in the docs to indicate that the model supports image embeddings.
> 
> Since the input format is different for this model, I suggest also adding a note to the documentation to explain how to input embeddings for Qwen2-VL.

Thanks for your tips.
Specifically, we are conducting a trial run of Qwen2VL with the new input format, as the model utilizes additional parameters for calculating positional encoding.
I think this feature could be extended in the future with minicpmv that also supports multiple images, such as video inputs.
If you guys need it and  agree with this, I might be able to refactor all.

whyiug commented:
While it is a rare use case and probably inconsequential to most users, it is a great enhancement to our production environment.

DarkLight1337 commented:
> > > You should also update the Supported Models page in the docs to indicate that the model supports image embeddings.
> > 
> > Since the input format is different for this model, I suggest also adding a note to the documentation to explain how to input embeddings for Qwen2-VL.
> 
> Thanks for your tips.
> Specifically, we are conducting a trial run of Qwen2VL with the new input format, as the model utilizes additional parameters for calculating positional encoding.

Thanks, the example looks good.

> I think this feature could be extended in the future with minicpmv that also supports multiple images, such as video inputs.
> If you guys need it and  agree with this, I might be able to refactor all.

@alex-jw-brooks are you already working on dynamic options to the input mapper?

DarkLight1337 commented:
I tested Qwen2-VL and can verify that the model still works on regular image inputs. Assuming that you have already tested Qwen2-VL to work with embedding inputs, let's merge this first - follow-up work can be done in another PR.

alex-jw-brooks commented:
Hey @DarkLight1337 - yup, I am working on it and should have a PR for passing dynamic processor / mapper options up in the next week or so!

focusunsink commented:
This is very needed
